react-native/react/actions/login2.js

```diff
@@ -277,25 +163,302 @@ export function logout () {
+  return (dispatch, getState) => {
+    const provisionMethod = enums.provisionUi.ProvisionMethod.device
+    startLoginFlow(dispatch, getState, provisionMethod, null, null, Constants.actionRegisteredWithExistingDevice)
+  }
+}
+
+export function registerWithUserPass () {
+  return (dispatch, getState) => {
+    const title = 'Registering with your Keybase passphrase'
+    const subTitle = 'lorem ipsum lorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsum'
+    const provisionMethod = enums.provisionUi.ProvisionMethod.passphrase
+    startLoginFlow(dispatch, getState, provisionMethod, title, subTitle, Constants.actionRegisteredWithUserPass)
+  }
+}
+
+export function registerWithPaperKey () {
+  return (dispatch, getState) => {
+    const title = 'Registering with your paperkey requires your username'
+    const subTitle = 'Different lorem ipsum lorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsum'
+    const provisionMethod = enums.provisionUi.ProvisionMethod.paperKey
+    startLoginFlow(dispatch, getState, provisionMethod, title, subTitle, Constants.actionRegisteredWithPaperKey)
+  }
+}
+
+function startLoginFlow (dispatch, getState, provisionMethod, userPassTitle, userPassSubtitle, successType) {
+  const mobile = true // TODO desktop also
+  const deviceType = mobile ? 'mobile' : 'desktop'
+  const incomingMap = makeKex2IncomingMap(dispatch, getState, provisionMethod, userPassTitle, userPassSubtitle, successType)
+
+  engine.rpc('login.login', {deviceType}, incomingMap, (error, response) => {
+    dispatch({
+      type: successType,
+      error: !!error,
+      payload: error || null
+    })
+
+    dispatch(navigateTo([]))
+    dispatch(switchTab(DEVICES_TAB))
+  })
+}
+
+export function addANewDevice () {
+  return (dispatch, getState) => {
+    const provisionMethod = enums.provisionUi.ProvisionMethod.device
+    const userPassTitle = 'Registering a new device requires your Keybase username and passphrase'
+    const userPassSubtitle = 'reggy lorem ipsum lorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsum'
+    const successType = null
+    const incomingMap = makeKex2IncomingMap(dispatch, getState, provisionMethod, userPassTitle, userPassSubtitle, successType)
+
+    engine.rpc('device.deviceAdd', {}, incomingMap, (error, response) => {
+      console.log(error)
+    })
+  }
+}
+
+function makeKex2IncomingMap (dispatch, getState, provisionMethod, userPassTitle, userPassSubtitle, successType) {
+  return {
+    'keybase.1.provisionUi.chooseProvisioningMethod': (param, response) => {
+      response.result(provisionMethod)
+    },
+    // TODO remove this when KEX2 supports not asking for this
+    'keybase.1.loginUi.getEmailOrUsername': (param, response) => {
+      const { username } = getState().login2.userPass
+      if (!username) {
+        dispatch(askForUserPass(userPassTitle, userPassSubtitle, () => {
```

Comment:
i have mixed feelings on this. in a sense i agree, but in another sense i think its ok since really the component side is dumb and the actions / store are smart so it makes sense for the routing to also happen there (especially since this is driven by the backend and not the user's choice). ultimately what happens is if we do a callback, the component will just need to be configured to know what to do next (which will come from an action) then it'll dispatch and action, which just does the routing anyways, so it does this round trip with a lot of bookkeeping , none of which originates from the view itself.